File: desktop/server/fetch_url.py
"""本機 sidecar 的 URL 抓取:server-side fetch,繞過瀏覽器 CORS。

合約與雲端 /api/fetch-url 一致(移植自 purism-ev-bot
services/utility_service.py 的 async/httpx 版)。
"""

import logging
import mimetypes
import os
from dataclasses import dataclass
from typing import Optional
from urllib.parse import unquote, urlparse

import httpx

logger = logging.getLogger(__name__)

# 常見 MIME → 副檔名:優先用顯式對映(避免 mimetypes.guess_extension 跨平台
# 結果不確定),未命中再回退 mimetypes。
_MIME_EXTENSION_MAP = {
    "application/pdf": ".pdf",
    "text/html": ".html",
    "text/plain": ".txt",
    "application/json": ".json",
    "image/jpeg": ".jpg",
    "image/png": ".png",
    "image/gif": ".gif",
    "application/epub+zip": ".epub",
}


def _guess_extension(content_type: str) -> Optional[str]:
    mime = content_type.split(";")[0].strip().lower()
    return _MIME_EXTENSION_MAP.get(mime) or mimetypes.guess_extension(mime)


@dataclass
class FetchResult:
    content: bytes
    content_type: str
    final_url: str
    redirect_count: int
    filename: str
    file_extension: str
    content_disposition: str


class FetchError(Exception):
    def __init__(self, message: str, status_code: int = 500):
        self.message = message
        self.status_code = status_code
        super().__init__(self.message)


async def fetch_url_content_async(
    url: str,
    follow_redirects: bool = True,
    max_redirects: int = 10,
    timeout: int = 30,
    client: Optional[httpx.AsyncClient] = None,
) -> FetchResult:
    if not url.startswith(("http://", "https://")):
        raise FetchError("URL 必須以 http:// 或 https:// 開頭", status_code=400)

    should_close_client = False
    if client is None:
        client = httpx.AsyncClient(
            timeout=httpx.Timeout(float(timeout)),
            follow_redirects=follow_redirects,
            max_redirects=max_redirects,
        )
        should_close_client = True

    try:
        response = await client.get(url)

        if response.status_code == 404:
            raise FetchError("找不到指定的資源", status_code=404)
        elif response.status_code == 429:
            raise FetchError("請求過於頻繁,請稍後再試", status_code=429)
        elif response.status_code >= 400:
            raise FetchError(
                f"HTTP 錯誤: {response.status_code}",
                status_code=response.status_code,
            )

        redirect_count = len(response.history)
        final_url = str(response.url)
        content_type = response.headers.get(
            "Content-Type", "application/octet-stream"
        )
        content = response.content

        parsed_url = urlparse(final_url)
        url_path = unquote(parsed_url.path)
        filename = os.path.basename(url_path) or "downloaded_file"

        if "." not in filename or filename.endswith(".php"):
            extension = _guess_extension(content_type)
            if extension:
                if filename.endswith(".php"):
                    filename = filename.rsplit(".", 1)[0] + extension
                else:
                    filename = filename + extension

        file_extension = os.path.splitext(filename)[1] if "." in filename else ""

        content_disposition = response.headers.get("Content-Disposition")
        if not content_disposition:
            content_disposition = f'inline; filename="{filename}"'

        return FetchResult(
            content=content,
            content_type=content_type,
            final_url=final_url,
            redirect_count=redirect_count,
            filename=filename,
            file_extension=file_extension,
            content_disposition=content_disposition,
        )

    except httpx.TimeoutException as e:
        raise FetchError("請求超時", status_code=408) from e
    except httpx.TooManyRedirects as e:
        raise FetchError(
            f"重定向次數超過限制 ({max_redirects} 次)", status_code=500
        ) from e
    except httpx.ConnectError as e:
        raise FetchError(f"無法連線到目標伺服器: {str(e)}", status_code=500) from e
    except httpx.RequestError as e:
        raise FetchError(f"抓取失敗: {str(e)}", status_code=500) from e
    except FetchError:
        raise
    except Exception as e:
        logger.error("URL 抓取未預期錯誤: %s", str(e), exc_info=True)
        raise FetchError(f"伺服器錯誤: {str(e)}", status_code=500) from e
    finally:
        if should_close_client:
            await client.aclose()

File: server/src/lib/fetchUrl.ts
import { lookup } from "node:dns";
import { isIP } from "node:net";
import { basename, extname } from "node:path";
import ipaddr from "ipaddr.js";
import { Agent, fetch as undiciFetch } from "undici";
import { ApiError, readLimited } from "./api";

const MAX_BYTES = 4 * 1024 * 1024;
const EXTENSIONS: Record<string, string> = {
  "application/pdf": ".pdf", "text/html": ".html", "text/plain": ".txt",
  "application/json": ".json", "image/jpeg": ".jpg", "image/png": ".png",
  "image/gif": ".gif", "application/epub+zip": ".epub",
};

export function isPublicAddress(address: string): boolean {
  try { return ipaddr.process(address).range() === "unicast"; } catch { return false; }
}

export function parseTarget(input: string): URL {
  let url: URL;
  try { url = new URL(input); } catch { throw new ApiError("無效的 URL", 400); }
  if (!["http:", "https:"].includes(url.protocol) || url.username || url.password) throw new ApiError("URL 必須使用 HTTP(S) 且不得包含帳密", 400);
  const hostname = url.hostname.replace(/^\[|\]$/g, "");
  if (hostname === "localhost" || hostname.endsWith(".localhost") || (isIP(hostname) && !isPublicAddress(hostname))) throw new ApiError("只允許抓取公開網路資源", 400);
  return url;
}

// Validate the actual socket lookup, not a separate DNS preflight; redirects use
// the same dispatcher so DNS changes cannot bypass the public-address boundary.
function publicAgent(): Agent {
  return new Agent({ connect: { lookup(hostname, options, callback) {
    lookup(hostname, { family: options.family, all: true }, (error, addresses) => {
      if (error) return callback(error, [], undefined);
      if (!addresses.length || addresses.some(({ address }) => !isPublicAddress(address))) {
        return callback(new Error("Non-public address"), [], undefined);
      }
      if (options.all) callback(null, addresses);
      else callback(null, addresses[0].address, addresses[0].family);
    });
  } } });
}

export function fetchOptions(params: URLSearchParams) {
  const url = params.get("url");
  if (!url) throw new ApiError("缺少 url", 422);
  const bool = (params.get("follow_redirects") ?? "true").toLowerCase();
  if (!["true", "false", "1", "0", "yes", "no", "on", "off", "t", "f", "y", "n"].includes(bool)) throw new ApiError("follow_redirects 必須是布林值", 422);
  const integer = (name: string, fallback: number, max: number) => {
    const raw = params.get(name) ?? String(fallback);
    if (!/^\d+$/.test(raw) || Number(raw) < 1 || Number(raw) > max) throw new ApiError(`${name} 必須介於 1 到 ${max}`, 422);
    return Number(raw);
  };
  return { url: parseTarget(url), follow: ["true", "1", "yes", "on", "t", "y"].includes(bool), redirects: integer("max_redirects", 10, 30), timeout: integer("timeout", 30, 120) };
}

export async function fetchUrl(request: Request, headers: Headers): Promise<Response> {
  const options = fetchOptions(new URL(request.url).searchParams);
  const agent = publicAgent();
  const deadline = AbortSignal.timeout(options.timeout * 1000);
  const signal = AbortSignal.any([request.signal, deadline]);
  let url = options.url;
  let count = 0;
  try {
    while (true) {
      const upstream = await undiciFetch(url, { dispatcher: agent, redirect: "manual", signal });
      const location = upstream.headers.get("location");
      if (options.follow && location && [301, 302, 303, 307, 308].includes(upstream.status)) {
        await upstream.body?.cancel();
        if (count >= options.redirects) throw new ApiError(`重定向次數超過限制 (${options.redirects} 次)`, 500);
        url = parseTarget(new URL(location, url).href);
        count++;
        continue;
      }
      if (upstream.status >= 400) {
        await upstream.body?.cancel();
        throw new ApiError(upstream.status === 404 ? "找不到指定的資源" : upstream.status === 429 ? "請求過於頻繁，請稍後再試" : `HTTP 錯誤: ${upstream.status}`, upstream.status);
      }
      const content = await readLimited(upstream as unknown as Response, MAX_BYTES);
      const type = upstream.headers.get("content-type") || "application/octet-stream";
      let filename = basename(decodeURIComponent(url.pathname)) || "downloaded_file";
      if (!filename.includes(".") || filename.endsWith(".php")) {
        const extension = EXTENSIONS[type.split(";")[0].trim().toLowerCase()];
        if (extension) filename = filename.replace(/\.php$/, "") + extension;
      }
      headers.set("Content-Type", type);
      headers.set("Content-Length", String(content.length));
      headers.set("X-Final-URL", url.href);
      headers.set("X-Redirect-Count", String(count));
      headers.set("X-File-Extension", extname(filename).replace(/[^\x20-\x7e]/g, ""));
      const safeName = filename.replace(/[^\x20-\x7e]|["\\]/g, "_");
      headers.set("Content-Disposition", upstream.headers.get("content-disposition") || `inline; filename="${safeName}"; filename*=UTF-8''${encodeURIComponent(filename)}`);
      headers.set("X-Content-Type-Options", "nosniff");
      headers.set("Content-Security-Policy", "sandbox; default-src 'none'");
      return new Response(content, { headers });
    }
  } catch (error) {
    if (error instanceof ApiError) throw error;
    if (deadline.aborted) throw new ApiError("請求超時", 408);
    throw new ApiError("無法抓取目標伺服器的資源", 500);
  } finally {
    await agent.destroy();
  }
}
